_posts/2016-03-10-information-hiding.md

```diff
@@ -0,0 +1,334 @@
+---
+layout: post
+title: Information hiding, enforced
+
+meta:
+  nav: blog
+  author: adelbertc
+  pygments: true
+---
+
+Code should be reusable. An expression traversing a data structure
+shouldn't be written multiple times, it should be pulled out into a
+generic traversal function. At a larger scale, a random number generator
+shouldn't be written multiple times, but rather pulled out into a
+module that can be used by others.
+
+It is important that such abstractions must be done carefully.
+Often times a type is visible to the caller, and if the type
+is not handled carefully the abstraction can leak.
+
+For example, a set with fast random indexing (useful for random
+walks on a graph) can be implemented with a sorted `Vector`.
+However, if the `Vector` type is
+leaked, the user can use this knowledge to violate the invariant.
+
+```scala
+import scala.annotation.tailrec
+
+def binarySearch(i: Int, repr: Vector[Int]): (Boolean, Int) = {
```

Comment:
It needs the actual implementation so Tut can generate the (commented)
output - I just commented out the implementation after Tut generation since
I do think it gets in the way of the more important stuff
On Mar 12, 2016 14:38, "Lars Hupel" <notifications@github.com> wrote:

> In _posts/2016-03-10-information-hiding.md
> <https://github.com/typelevel/typelevel.github.com/pull/93#discussion_r55928886>
> :
>
> > +shouldn't be written multiple times, but rather pulled out into a
> > +module that can be used by others.
> > +
> > +It is important that such abstractions must be done carefully.
> > +Often times a type is visible to the caller, and if the type
> > +is not handled carefully the abstraction can leak.
> > +
> > +For example, a set with fast random indexing (useful for random
> > +walks on a graph) can be implemented with a sorted `Vector`.
> > +However, if the `Vector` type is
> > +leaked, the user can use this knowledge to violate the invariant.
> > +
> > +```scala
> > +import scala.annotation.tailrec
> > +
> > +def binarySearch(i: Int, repr: Vector[Int]): (Boolean, Int) = {
>
> You could replace it by ??? [image: :zap:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/typelevel/typelevel.github.com/pull/93/files#r55928886>
> .
>